numpy/dtypes.pyi

```diff
@@ -1,43 +1,641 @@
+import sys
+from typing import (
+    Any,
+    Final,
+    Generic,
+    Literal as L,
+    NoReturn,
+    TypeAlias,
+    TypeVar,
+    final,
+)
+
 import numpy as np
 
+if sys.version_info >= (3, 11):
+    from typing import LiteralString
+else:
+    LiteralString: TypeAlias = str
+
+
+__all__ = [
+    'BoolDType',
+    'Int8DType',
+    'ByteDType',
+    'UInt8DType',
+    'UByteDType',
+    'Int16DType',
+    'ShortDType',
+    'UInt16DType',
+    'UShortDType',
+    'Int32DType',
+    'IntDType',
+    'UInt32DType',
+    'UIntDType',
+    'Int64DType',
+    'LongDType',
+    'UInt64DType',
+    'ULongDType',
+    'LongLongDType',
+    'ULongLongDType',
+    'Float16DType',
+    'Float32DType',
+    'Float64DType',
+    'LongDoubleDType',
+    'Complex64DType',
+    'Complex128DType',
+    'CLongDoubleDType',
+    'ObjectDType',
+    'BytesDType',
+    'StrDType',
+    'VoidDType',
+    'DateTime64DType',
+    'TimeDelta64DType',
+    'StringDType',
+]
+
+_SelfType = TypeVar("_SelfType", bound=object)
+_SCT = TypeVar("_SCT", bound=np.generic, covariant=True)
+
+# Helper base classes (typing-only)
+
+class _BaseDType(Generic[_SCT], np.dtype[_SCT]):  # type: ignore[misc]
+    names: None  # pyright: ignore[reportIncompatibleVariableOverride]
+    def __new__(cls: type[_SelfType], /) -> _SelfType: ...
+    def __getitem__(self, key: Any, /) -> NoReturn: ...
+    @property
+    def base(self) -> np.dtype[_SCT]: ...
+    @property
+    def fields(self) -> None: ...
+    @property
+    def isalignedstruct(self) -> L[False]: ...
+    @property
+    def isnative(self) -> L[True]: ...
+    @property
+    def metadata(self) -> None: ...
+    @property
+    def ndim(self) -> L[0]: ...
+    @property
+    def shape(self) -> tuple[()]: ...
+    @property
+    def subdtype(self) -> None: ...
+
+class _BuiltinDType(Generic[_SCT], _BaseDType[_SCT]):
+    @property
+    def flags(self) -> L[0]: ...
+    @property
+    def hasobject(self) -> L[False]: ...
+    @property
+    def isbuiltin(self) -> L[1]: ...
```

Comment:
It's not often that you understand less and less of something, the more you know about it, like with quantum mechanics 😅

Reply:
Its just weird:
* 1 is a *prototype* instance, i.e. the default singleton instance
* 2 is a non builtin prototype instance, IIRC.
* 0 is everything else.

And now, you will have to forget it again, because its just baggage.

Reply:
That's quite a bit different from what the docs say about it 🤔: https://numpy.org/doc/stable/reference/generated/numpy.dtype.isbuiltin.html
Either way, your explanation makes a lot more sense.